File: 735. Asteroid Collision.py
class Solution:
    def asteroidCollision(self, a: List[int]) -> List[int]:
        '''
        using a stack for keeping track of curr as and new as,
        note for the writing style of while / else
        '''
        stack = []
        for new in a:
            while stack and new < 0 < stack[-1]:
                if stack[-1] < -new:
                    stack.pop()
                    continue
                elif stack[-1] == -new:
                    stack.pop()
                break
            else:
                stack.append(new)
        return stack

File: 735. Asteroid Collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i : asteroids) {
            if (i < 0) {
                boolean isBreak = false;
                while (!stack.isEmpty() && stack.peek() > 0) {
                    if (i < -1 * stack.peek()) {
                        stack.pop();
                    } else {
                        if (i == -1 * stack.peek()) {
                            stack.pop();
                        }
                        isBreak = true;
                        break;
                    }
                }
                if (!isBreak) {
                    stack.push(i);
                }
            } else {
                stack.push(i);
            }
        }
        int[] ans = new int[stack.size()];
        for (int i=0; i<ans.length; i++) {
            ans[i] = stack.pollLast();
        }
        return ans;
    }
}